File: ops/built-in/op_tiling/rnn_gen_mask_v2.cc
/**
 * Copyright (c) Huawei Technologies Co., Ltd. 2021. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <string>
#include <math.h>
#include <iostream>
#include <nlohmann/json.hpp>
#include "op_tiling.h"
#include "op_log.h"
#include "error_log.h"

namespace optiling{
  struct RnnGenMaskV2TilingParams
  {
    int32_t cal_mode;
    int32_t core_used;
    int32_t batch_size;
    int32_t num_step;
    int32_t rounds;
    int32_t batch_num_per_aicore;
    int32_t batch_tail;
  };

  void InitTilingParams(RnnGenMaskV2TilingParams &params)
  {
    OP_LOGD("InitTilingParams is running");
    params.cal_mode = 0;
    params.core_used = 0;
    params.batch_size = 0;
    params.num_step = 0;
    params.rounds = 0;
    params.batch_num_per_aicore = 0;
    params.batch_tail = 0;
  }

  bool GetCompileInfoV2(const std::string &op_type, const nlohmann::json &op_compile_info,
                        int32_t &aicore_num, int32_t &block)
  {
    OP_LOGD("GetCompileInfoV2 is running");
    using namespace nlohmann;
    auto all_vars = op_compile_info["vars"];
    if (all_vars.count("available_aicore_num") == 0)
    {
      VECTOR_INNER_ERR_REPORT_TILIING("RnnGenMaskV2Tiling", "GetCompileInfoV2, get available_aicore_num  error");
      return false;
    }
    aicore_num = all_vars["available_aicore_num"].get<std::int32_t>();
    if (all_vars.count("block") == 0)
    {
      VECTOR_INNER_ERR_REPORT_TILIING("RnnGenMaskV2Tiling", "GetCompileInfoV2, get block error");
      return false;
    }
    block = all_vars["block"].get<std::int32_t>();
    return true;
  }

  int32_t CalTilingModeV2(std::vector<int64_t> x_shape)
  {
    OP_LOGD("CalTilingModeV2 is running");
    int32_t tiling_mode = 0;
    auto batch_size = x_shape[1];
    if (batch_size >= 1)
    {
      tiling_mode = 1;
    }
    return tiling_mode;
  }
    
  static void CalCoreInfo(RnnGenMaskV2TilingParams &tiling_params, int32_t core_num, 
		          const std::vector<int64_t> & x_shape){
    OP_LOGD("CalCoreInfo is running");
    int32_t batch = x_shape[1];
    int32_t num_step = x_shape[0];
    auto rounds = batch * num_step;
    int32_t batch_num_per_aicore = 0;
    int32_t core_used = 0;
    int32_t batch_tail = 0;

    if (rounds > core_num)
    {
      core_used = core_num;
    }
    else
    {
      core_used = rounds;
    }
    OP_TILING_CHECK(core_num == 0, VECTOR_INNER_ERR_REPORT_TILIING("rnn_gen_mask_v2",
      "core_num = 0 is not support"), return);
    batch_num_per_aicore = rounds / core_used;
    batch_tail = rounds % core_used;

    tiling_params.batch_num_per_aicore = batch_num_per_aicore;
    tiling_params.core_used = core_used;
    tiling_params.batch_tail = batch_tail;
  }

  void CalRunningInfo(RnnGenMaskV2TilingParams &tiling_params, int32_t core_num, int32_t block,
                      std::vector<int64_t> & x_shape)
  {
    OP_LOGD("CalRunningInfo is running");
    OP_TILING_CHECK(block == 0, VECTOR_INNER_ERR_REPORT_TILIING("rnn_gen_mask_v2",
      "block = 0 is not support"), return);
    int32_t batch_size = x_shape[1];
    int32_t num_step = x_shape[0];
    int32_t rounds = batch_size * num_step;

    tiling_params.cal_mode = CalTilingModeV2(x_shape);
    tiling_params.batch_size = batch_size;
    tiling_params.num_step = num_step;
    tiling_params.rounds = rounds;
    CalCoreInfo(tiling_params, core_num, x_shape);
  }

  void SetRunningInfo(const RnnGenMaskV2TilingParams &tiling_params, OpRunInfo &run_info)
  {
    OP_LOGD("SetRunningInfo is running");
    ByteBufferPut(run_info.tiling_data, tiling_params.cal_mode);
    ByteBufferPut(run_info.tiling_data, tiling_params.core_used);
    ByteBufferPut(run_info.tiling_data, tiling_params.batch_size);
    ByteBufferPut(run_info.tiling_data, tiling_params.num_step);
    ByteBufferPut(run_info.tiling_data, tiling_params.rounds);
    ByteBufferPut(run_info.tiling_data, tiling_params.batch_num_per_aicore);
    ByteBufferPut(run_info.tiling_data, tiling_params.batch_tail);
  }

  void PrintTilingParams(const RnnGenMaskV2TilingParams &tiling_params)
  {
    OP_LOGD("PrintTilingParams is running");
    OP_LOGD("op [RnnGenMaskV2Tiling] : cal_mode=%d.", tiling_params.cal_mode);
    OP_LOGD("op [RnnGenMaskV2Tiling] : core_used=%d.", tiling_params.core_used);
    OP_LOGD("op [RnnGenMaskV2Tiling] : batch_size=%d.", tiling_params.batch_size);
    OP_LOGD("op [RnnGenMaskV2Tiling] : num_step=%d.", tiling_params.num_step);
    OP_LOGD("op [RnnGenMaskV2Tiling] : rounds=%d.", tiling_params.rounds);
    OP_LOGD("op [RnnGenMaskV2Tiling] : batch_num_per_aicore=%d.", tiling_params.batch_num_per_aicore);
    OP_LOGD("op [RnnGenMaskV2Tiling] : batch_tail=%d.", tiling_params.batch_tail);
  }

  bool RnnGenMaskV2Tiling(const std::string &op_type, const TeOpParas &op_paras,
                          const nlohmann::json &op_compile_info, OpRunInfo &run_info)
  {
    OP_LOGD("RnnGenMaskV2Tiling is running");
    int32_t core_num;
    int32_t block;
    bool get_compile_info = GetCompileInfoV2(op_type, op_compile_info, core_num, block);
    if (!get_compile_info)
    {
      VECTOR_INNER_ERR_REPORT_TILIING(op_type, "RnnGenMaskV2Tiling: GetCompileInfoV2 error.");
      return false;
    }

    RnnGenMaskV2TilingParams tiling_params;
    InitTilingParams(tiling_params);
    std::vector<int64_t> x_shape = op_paras.inputs[1].tensor[0].shape;
    CalRunningInfo(tiling_params, core_num, block, x_shape);
    SetRunningInfo(tiling_params, run_info);
    PrintTilingParams(tiling_params);

    run_info.block_dim = tiling_params.core_used;
    std::vector<int64_t> workspace={81920};
    run_info.workspaces = workspace;
    return true;
  }
  REGISTER_OP_TILING_FUNC_BUFFERED(RnnGenMaskV2, RnnGenMaskV2Tiling);
} // namespace optiling.

File: ops/built-in/tbe/impl/dynamic/rnn_gen_mask_v2.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Copyright 2021 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
"""
rnn_gen_mask
"""

from impl.util.platform_adapter import tik
from impl.util.platform_adapter import tbe_context
from impl.util.platform_adapter import register_operator
from impl.util.platform_adapter import tbe_platform as cce


# 'pylint: disable=too-few-public-methods
class Constant:
    """
    The class for constant
    """
    TILING_ARG_NUM = 32
    MAX_INT32 = 2**31 - 1
    SCALAR_TENSOR_SIZE = 32
    UB_AVALIB_SIZE = 10240


# 'pylint: disable=too-many-instance-attributes
class RnnGenMaskV2:
    """class for rnn_gen_mask"""

    # 'pylint: disable=too-many-arguments
    def __init__(self, seq_length, x, seq_mask, hidden_size, kernel_name):
        """__init__"""
        self.tik_instance = tik.Tik(tik.Dprofile())
        self.available_aicore_num = tik.Dprofile().get_aicore_num()

        self.kernel_name = kernel_name
        self.tiling_param_dtype = 'int32'
        self.dtype = seq_length.get("dtype").lower()
        self.dtype_b = seq_mask.get("dtype").lower()
        self.dtype_x = x.get("dtype").lower()

        if self.dtype_b == "float16":
            self.block = 16
        if self.dtype_b == "float32":
            self.block = 8

        self.seq_length_gm = self.tik_instance.Tensor(self.dtype, [Constant.MAX_INT32],
                                                      name="seq_length_gm",
                                                      scope=tik.scope_gm)
        self.seq_mask_gm = self.tik_instance.Tensor(self.dtype_b, [Constant.MAX_INT32],
                                                    name="seq_mask_gm",
                                                    scope=tik.scope_gm)
        self.temp_gm = self.tik_instance.Tensor(self.dtype_b, [Constant.UB_AVALIB_SIZE],
                                                name="temp_gm",
                                                scope=tik.scope_gm,
                                                is_workspace=True)
        self.x_gm = self.tik_instance.Tensor(self.dtype_x, [Constant.MAX_INT32], name="x_gm", scope=tik.scope_gm)
        self.tiling_gm = self.tik_instance.Tensor('int32', (Constant.TILING_ARG_NUM,),
                                                  name="tiling_gm",
                                                  scope=tik.scope_gm)
        block_bite_size = 32
        self.tiling_dtype = 'int32'
        self.dtype_bytes_size_tiling = cce.get_bit_len(self.tiling_dtype) // 8
        self.tiling_each_block = block_bite_size // self.dtype_bytes_size_tiling
        self.tiling_gm = self.tik_instance.Tensor(self.tiling_dtype, (Constant.TILING_ARG_NUM,),
                                                  name='tiling_gm',
                                                  scope=tik.scope_gm)

        self.hidden_size = hidden_size
        self.hidden_size_block = (self.hidden_size + self.block - 1) // self.block * self.block
        self.repeat = self.hidden_size_block // self.block

        self.cal_mode = None
        self.core_used = None
        self.batch_size = None
        self.num_step = None
        self.rounds = None
        self.batch_num_per_aicore = None
        self.batch_tail = None

    def get_tiling_args(self):
        """get_tiling_args"""
        tiling_ub = self.tik_instance.Tensor(self.tiling_dtype, (Constant.TILING_ARG_NUM,),
                                             name='tiling_ub',
                                             scope=tik.scope_ubuf)
        self.tik_instance.data_move(tiling_ub, self.tiling_gm, 0, 1,
                                    Constant.SCALAR_TENSOR_SIZE // self.tiling_each_block, 0, 0)

        self.cal_mode = self.tik_instance.Scalar(self.tiling_param_dtype, name='cal_mode')
        self.core_used = self.tik_instance.Scalar(self.tiling_param_dtype, name='core_used')
        self.batch_size = self.tik_instance.Scalar(self.tiling_param_dtype, name='batch_size')
        self.num_step = self.tik_instance.Scalar(self.tiling_param_dtype, name='num_step')
        self.rounds = self.tik_instance.Scalar(self.tiling_param_dtype, name='rounds')
        self.batch_num_per_aicore = self.tik_instance.Scalar(self.tiling_param_dtype, name='batch_num_per_aicore')
        self.batch_tail = self.tik_instance.Scalar(self.tiling_param_dtype, name='batch_tail')

        self.cal_mode.set_as(tiling_ub[0])
        self.core_used.set_as(tiling_ub[1])
        self.batch_size.set_as(tiling_ub[2])
        self.num_step.set_as(tiling_ub[3])
        self.rounds.set_as(tiling_ub[4])
        self.batch_num_per_aicore.set_as(tiling_ub[5])
        self.batch_tail.set_as(tiling_ub[6])

    def compute_core(self, task_id):
        """compute_core"""
        temp_ub = self.tik_instance.Tensor("int32", [self.block], name="temp_ub", scope=tik.scope_ubuf)
        self.tik_instance.data_move(temp_ub, self.seq_length_gm[task_id % self.batch_size], 0, 1, 1, 0, 0)
        valid_len = self.tik_instance.Scalar("int32", init_value=temp_ub[0])
        curre_len = self.tik_instance.Scalar("int32", init_value=task_id // self.batch_size)

        mask_ub = self.tik_instance.Tensor(self.dtype_b, [Constant.UB_AVALIB_SIZE],
                                           name="mask_ub",
                                           scope=tik.scope_ubuf)
        with self.tik_instance.if_scope(valid_len > curre_len):
            self.tik_instance.vector_dup(self.block, mask_ub, 1, self.repeat, 1, 1)
        with self.tik_instance.else_scope():
            self.tik_instance.vector_dup(self.block, mask_ub, 0, self.repeat, 1, 1)

        with self.tik_instance.if_scope(self.hidden_size != self.hidden_size_block):
            self.tik_instance.data_move(self.temp_gm[task_id * self.hidden_size_block], mask_ub, 0, 1, self.repeat, 0,
                                        0)
        with self.tik_instance.else_scope():
            self.tik_instance.data_move(self.seq_mask_gm[task_id * self.hidden_size], mask_ub, 0, 1, self.repeat, 0, 0)

    def data_tune(self):
        """data_tune"""
        with self.tik_instance.for_range(0, self.rounds, thread_num=2) as i:
            temp_ub = self.tik_instance.Tensor(self.dtype_x, [Constant.UB_AVALIB_SIZE],
                                               name="temp_ub",
                                               scope=tik.scope_ubuf)
            self.tik_instance.data_move(temp_ub, self.temp_gm[i * self.hidden_size_block], 0, 1, self.repeat, 0, 0)
            self.tik_instance.data_move(self.seq_mask_gm[i * self.hidden_size], temp_ub, 0, 1, self.repeat, 0, 0)

    def rnn_gen_mask_compute(self):
        """rnn_gen_mask_compute"""
        with self.tik_instance.for_range(0, self.available_aicore_num, block_num=self.available_aicore_num) as i:
            self.get_tiling_args()
            with self.tik_instance.if_scope(i < self.core_used):
                with self.tik_instance.for_range(0, self.batch_num_per_aicore) as k:
                    self.compute_core(i + k * self.core_used)
                with self.tik_instance.if_scope(i < self.batch_tail):
                    self.compute_core(i + self.batch_num_per_aicore * self.core_used)

        self.get_tiling_args()
        with self.tik_instance.if_scope(self.hidden_size != self.hidden_size_block):
            self.data_tune()

        self.tik_instance.BuildCCE(kernel_name=self.kernel_name,
                                   inputs=[self.seq_length_gm, self.x_gm],
                                   outputs=[self.seq_mask_gm],
                                   flowtable=[self.tiling_gm])
        tbe_context.get_context().add_compile_info('vars', {
            'available_aicore_num': self.available_aicore_num,
            'block': self.block
        })
        return self.tik_instance


# 'pylint: disable=invalid-name
@register_operator('rnn_gen_mask_v2')
def rnn_gen_mask_v2(seq_length, x, seq_mask, hidden_size, kernel_name="rnn_gen_mask_v2"):
    """
    Function: rnn_gen_mask_v2.
    Modify : 2021-04-22

    Init base parameters
    Parameters
    ----------
    input(seq_length): dict
        data of input
    input(x): dict
        data of input
    input(b): dict
        data of input
    output(seq_mask): dict
        data of output
    kernel_name: str
        the name of the operator
    ----------
    """
    op_obj = RnnGenMaskV2(seq_length, x, seq_mask, hidden_size, kernel_name)
    return op_obj.rnn_gen_mask_compute()
